pulp_container/app/token_verification.py

```diff
@@ -127,10 +128,14 @@ def has_permission(self, request, view):
         """
         Decide upon permission based on token
         """
+        decoded_token = request.auth
+        if decoded_token is True:
+            return True
+
         try:
-            decoded_token = request.auth
-            return decoded_token is True or _contains_accessible_actions(
-                decoded_token, *_access_scope(request)
-            )
+            if _contains_accessible_actions(decoded_token, *_access_scope(request)):
+                return True
         except Exception:
-            return False
+            raise NotAuthenticated()
```

Comment:
No code here?

Also the examples in https://www.django-rest-framework.org/api-guide/permissions/ all return boolean.
And the docs talk about a `PermissionDenied` exception. (It has a `code` attribute too.)

Reply:
No error code means that a user did not provide any authentication credentials. In this case, dockerhub returns an "UNAUTHORIZED" response without the error code. I tried to employ the same pattern in Pulp.

Speaking of `PermissionDenied`, all raised exceptions are handled by the method `handle_exception()`:https://github.com/pulp/pulp_container/blob/ca86159b39f110087abc7eac9f24ff5233460f49/pulp_container/app/registry_api.py#L237

Here, the authenticate header is built after invoking `super().handle_exception(exc)`: https://github.com/encode/django-rest-framework/blob/d635bc9c71b6cf84d137a68610ae2e628f8b62b3/rest_framework/views.py#L456. Since we already construct the authenticate header by ourselves, with my changes we just append the error code to it only when the exception `AuthenticationFailed` with a code is raised.

On the other hand, the exception `PermissionDenied` does not result into calling the method `authenticate_header()`: https://github.com/pulp/pulp_container/blob/ca86159b39f110087abc7eac9f24ff5233460f49/pulp_container/app/token_verification.py#L100

We would be required to parse the request manually, as they do in `django-rest-framework` and make the method `authenticate_header()` to be accessible globally. But then we will return "DENIED" in the response message, as opposed to "UNAUTHORIZED" in dockerhub.

Reply:
i agree we should return 401 and not 403

Reply:
What needs to happen anyway that we come to this point and raise this exception?
Is this a case of "invalid-token" or "no-token"?

Reply:
`NotAuthenticated` means that a user did not provide any authentication credentials.

Reply:
In that case, `request.auth` should be None, right? That should work.